src/LeafInputSignature.cpp

```diff
@@ -0,0 +1,403 @@
+                            HeaderValidationError error;
+                            error.runtimeNode = runtimeNodeWithScanner;
+                            error.included_file = fileName;
+                            BufferAppendOne(&illegalIncludesToReport, &thread_state->m_LocalHeap, error);
+                        }
+                    }
+                }
+                break;
+
+            case NodeBuildResult::kRanSuccesfully:
+                HashSetWalk(&runtimeNodeWithScanner->m_ImplicitInputs, [&](uint32_t, uint32_t hash, const char *filename) {
+                    if (!IsGeneratedOrIsLeafInput(hash, filename))
+                    {
+                        HeaderValidationError error;
+                        error.runtimeNode = runtimeNodeWithScanner;
+                        error.included_file = filename;
+                        BufferAppendOne(&illegalIncludesToReport, &thread_state->m_LocalHeap, error);
+                    }
+                });
+                break;
+
+            default:
+                Croak("Unexpected build node result of dependent node while verifying headers");
+        }
+    }
+
+    HashSetDestroy(&alreadyFound);
+
+    if (illegalIncludesToReport.m_Size > 0)
+    {
+        auto error = illegalIncludesToReport[0];
+
+        ExecResult result = {0, false};
+        result.m_ReturnCode = 1;
+
+        InitOutputBuffer(&result.m_OutputBuffer, &thread_state->m_LocalHeap);
+
+        {
+            MemAllocLinearScope scope(&thread_state->m_ScratchAlloc);
+
+            const char* formatString =
+            "This node '%s' is marked as leaf input cacheable.\n"
+            "%d implicit input files have been discovered that were not statically known ahead of time.\n"
+            "For example:\n"
+            "The dependency node '%s' ends up including '%s'.\n"
+            "This usually only happens if a file is only included by generated files.\n"
+            "You can fix this by either:\n"
+            "- Making '%s' not be included by the generated files including it currently.\n"
+            "- By ensuring that this build has a non-generated file that also includes '%s'.\n"
+            "- Adding this headerfile to the .Sources property of your NativeProgram.\n";
+
+            const int kStoragePerInclude = kMaxPathLength + 50;
+            size_t errorStorageSize = strlen(formatString)*2 + illegalIncludesToReport.m_Size * kStoragePerInclude;
+            char* errorStorage = LinearAllocateArray<char>(&thread_state->m_ScratchAlloc,errorStorageSize);
+
+            char* endPtr = errorStorage + errorStorageSize;
+
+            int written = snprintf(errorStorage,errorStorageSize, formatString,
+                node->m_DagNode->m_Annotation.Get(),
+                illegalIncludesToReport.m_Size,
+                error.runtimeNode->m_DagNode->m_Annotation.Get(),
+                error.included_file,
+                error.included_file,
+                error.included_file
```

Comment:
nitpick: It might be a bit of overkill to mention the filename several times. Maybe just mention it once and then refer to it as "the header file" ?

Reply:
It's a very complicated error message, that i think is of higher quality by specifying very specifically what the problem is.  that just requires writing out that included file several times.
If anything we need to make this error even better because richard was still confused when he hit it the first time